### DISCUSSION TITLE
Error wont go away (lazy.nvim luarocks)

### DISCUSSION THREAD

**Discussion by d7omdev** [3rd/image.nvim#239] [General]
I keep getting this error whenever I open neovim although the plugin is working 

![image](https://github.com/3rd/image.nvim/assets/61534551/1f0a0c48-92b3-449e-90fa-e119ae76142d)

**Comment by onurcanbektas** [comment]
Same here

**Comment by 3rd** [comment]
hey @kevinm6 do you know how we could avoid this? i think we don't have to force users to have a global lua 5.1 install (that fixes it) since it will be loaded by luajit anyway right?

**Comment by 3rd** [comment]
Pinned this and disabled rockspec until we find a workaround @abdulrahmanDev1 @kevinm6
There is a discussion about this on Reddit, I'm not that familiar with rocks, hopefully we can find the right way to do things soon: https://www.reddit.com/r/neovim/comments/1dng1d6/comment/lc2t5uh/

**Comment by kevinm6** [comment]
Hi @3rd .. unfortunately for now it's [required](https://github.com/folke/lazy.nvim/issues/1570) from Lazy the Lua@5.1 ! 
A workaround is the one reported in the Lazy UI warnings (or **checkhealth**), in **Lazy** config set this

```lua
rocks = {
  hererocks = true
}
```

Another [pointer](https://github.com/folke/lazy.nvim/issues/1548#issuecomment-2189831348) where it was discussed

**Comment by 3rd** [comment]
What I think I'm missing is why is having a specific Lua version installed externally required when the end goal is pulling luarocks that will be executed with the LuaJIT bundled by Neovim anyway - I don't think I understand how it works. Is it for a build step on some of the rocks? Or is the external Lua installation used at runtime?

**Comment by kevinm6** [comment]
I think is due to some internals of **Lazy** that are really integrated with Neovim, since that's the one that needs Lua@5.1 and the way it loads the modules at runtime.

I don't think depends on something relative to **LuaJIT** or the rock build itself, since I'm capable of build it without **Lua5.1** and  in multiple ways outside NeoVim

**Comment by sebalfaro** [comment]
Hey guys @kevinm6 i'm stucked in the same point, so should i install globally lua 5.1 and also the the hererocks key on true, like this?

```
return {
  "3rd/image.nvim",
  dependencies = {
    "leafo/magick",
    "nvim-treesitter/nvim-treesitter",
    {
      "vhyrro/luarocks.nvim",
      opts = {
        rocks = {
          hererocks = true,
        },
      },
    },
  },
```

**Comment by 3rd** [comment]
I think if you install 5.1 globally it should just work, luarocks.nvim says it needs a global LuaJIT/Lua 5.1 installation.

**Comment by sebalfaro** [comment]
Thank you @3rd

**Comment by kapral18** [comment]
@3rd  luajit is installed with neovim when using `brew` because it's a core dependency. I have it now and I see the error about missing magick rock.

**Comment by kevinm6** [comment]
> Hey guys @kevinm6 i'm stucked in the same point, so should i install globally lua 5.1 and also the the hererocks key on true, like this?
> 
> 
> 
> ```lua
>   rocks = {
>     hererocks = true,
>   },
> ```

Sorry, maybe I wasn't clear.. the option `rocks.hererocks = true` is **Lazy**'s one (it installs another plugin that is a python script and manage the rocks and lua versions installs).
If you enable it, it works without having to install manually **Lua5.1** and make it available on `$PATH`.

There are other ways you can fix it, as reported from @3rd, like just install and make lua5.1 available on **$PATH** or if you stick with `luarocks.nvim`, just do it as before.

**Comment by kevinm6** [comment]
> @3rd  luajit is installed with neovim when using `brew` because it's a core dependency. I have it now and I see the error about missing magick rock.

@kapral18 
If you update `image.nvim` to the last commit, you should follow the **Installation** steps and requirements as before

**Comment by sebalfaro** [comment]
@kevinm6  @3rd working on my end without erros, i've added the ` hererocks` line that @kevinm6  has suggested. Thank you both guys

```lua
  rocks = {
    hererocks = true,
  },
```

**Comment by d7omdev** [comment]
Thank you guys it's fixed for me too after I added the hererocks line I tried to do it before but it didn't work out for me,
Thank you all.

**Comment by kevinm6** [comment]
> @kevinm6  @3rd working on my end without erros, i've added the `hererocks` line that @kevinm6  has suggested. Thank you both guys

@sebalfaro @abdulrahmanDev1 
Glad it helped. The amazing job is @3rd 's one, it seems now, my little contribution created more issues than benefits!

**Comment by 3rd** [comment]
@kevinm6 no no, your contribution is great, so Lazy requires Lua 5.1 installed correct?
Then we can re-enable the rockspec and document that requirement as well, would that be the best?

**Comment by sebalfaro** [comment]
@3rd at least in my case it wasn't necessary to install neither Lua 5.1 nor magick locally, just adding the `hererocks` line in the `lazy.lua` file

**Comment by kevinm6** [comment]
> @kevinm6 no no, your contribution is great, so Lazy requires Lua 5.1 installed correct?

Yes, **Lazy** requires lua5.1 for the **luarocks**, due to its integration. 

> Then we can re-enable the rockspec and document that requirement as well, would that be the best?

I think should be fine if it's okay for you!
We could try to make it really clear in the **README** and probably is the better choice to use so far, since is reported from **Lazy**'s developer

**Comment by joncrangle** [comment]
Thanks for everyone's help in this thread. Since I didn't have any other rocks installed in my config, I found that adding the `hererocks` rock to my `lazy` config wasn't enough. I needed to specify the magick dependency to initiate `hererocks` getting added as a plugin and the `magick` plugin bindings getting built. Sharing my setup (with a utility function I use to disable on Windows) in case anyone else runs into this and might find it helpful:
```lua
return {
  {
    '3rd/image.nvim',
    -- Utility function to disable on Windows system
    enabled = function()
      if vim.fn.has 'win32' == 1 then
        return false
      else
        return true
      end
    end,
    dependencies = {
      'leafo/magick',
    },
    opts = {
      -- ... image.nvim config
    },
  },
}
```

**Comment by kevinm6** [comment]
@joncrangle I think you can improve it and use **Lazy** `cond` to do not load a plugin if the condition isn't satisfied.

```lua
return {
{
   '3rd/image.nvim',
    -- Disable on Windows system
    cond = function()
       return vim.fn.has 'win32' ~= 1 
    end,
    dependencies = {
       'leafo/magick',
    },
    opts = {
       -- image.nvim config
    }
}
}
```

**Comment by NeelMishra** [comment]
Any updates on this issue?

**Comment by 3rd** [comment]
> Any updates on this issue?

What issue are you encountering?
You can opt out of lazy + rocks and install the magick rock manually.

**Comment by sankantsu** [comment]
Thank you for great plugin! I recently developed [plugin utilizing image.nvim](https://github.com/sankantsu/gin-diff-image.nvim).
image.nvim saved a lot of work for correctly handling scrolling/redrawing tasks!

I did some research about luarocks and lazy.nvim, and considering if we can re-enable rockspec support for lazy.nvim (which was introduced in https://github.com/3rd/image.nvim/pull/185).
I tested locally and the rockspec seems working without error when I set `opts.rocks.hererocks` (as already mentioned in https://github.com/3rd/image.nvim/issues/191#issuecomment-2214148798).
(NOTE: Even if `opts.rocks.hererocks` is not set and global Lua 5.4 (not Lua 5.1), image.nvim and magick build seems working, but `:checkhealth` of lazy.nvim complains. So I think it is not recommended configuration).

I wrote PR in my fork which should resolve this issue.
https://github.com/sankantsu/image.nvim/pull/1

If you would like, I will happily send PR to your repo.
Thank you!

**Comment by 3rd** [comment]
Hey, thanks awesome! Please open the PR <3

**Comment by paulwyszynski** [comment]
Sorry for bringing that up again, but for me it still does not work actually.

I followed the instructions in the Readme:

1. Added this to my zshrc file:
`export DYLD_FALLBACK_LIBRARY_PATH="$(brew --prefix)/lib:$DYLD_FALLBACK_LIBRARY_PATH"`
2. Added this to lazy.lua:
```
rocks = {
        hererocks = true,  -- recommended if you do not have global installation of Lua 5.1.
    },
```

3. And my image.lua file for loading the plugin:
```
  {
    "3rd/image.nvim",
    opts = {
      integrations = {
        markdown = {
          enabled = true,
          clear_in_insert_mode = true,
          only_render_image_at_cursor = true,
        },
      },
      html = {
        enabled = true,
      },
      css = {
        enabled = true,
      },
      max_width_window_percentage = 50,
      tmux_show_only_in_active_window = true,
    },
  },

```
This throws this error when I open nvim: `image.nvim: magick rock not found, please install it and restart your editor. Error: "loop or previous error loading module 'magick'`

MacOS Sequoia
Kitty: 0.37.0
Lazy: 11.14.2
luarocks (health-check):
```
luarocks ~
- checking `hererocks` installation
- you have some plugins that require `luarocks`:
    * `hererocks`
    * `image.nvim`
- OK {python3} `Python 3.9.6`
- OK {/Users/paule/.local/share/nvim/lazy-rocks/hererocks/bin/luarocks} `3.11.1`
- OK {/Users/paule/.local/share/nvim/lazy-rocks/hererocks/bin/lua} `Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio`
```

**Reply by paulwyszynski** [reply]
Ok it works now after installing pkg-config.
`brew install pkg-config`
Would be nice to add this to the installation process.

**Reply by 3rd** [reply]
you can also opt-out of using the rock and switch to the magick_cli backend
